### PR TITLE
Make client and server to resync active connections

### PR DESCRIPTION
**Pull request by aruiz14**
## Issue: [44576](https://github.com/rancher/rancher/issues/44576)

Relates to #68
Depends on #78

## Problem

remotedialer allows multiplexing connections between two peers using a single websocket connections by including a connection ID in the messages and using separate buffers. The protocol specifies different message types for different actions (`Connect`, `Data`, `Error`, `Pause`, `Resume`, etc.). In particular, the `Error` type is used to communicate the other end that a certain connection must be closed. However, depending on the cause of the original error, this message may never be successfully transmitted, as the sender will give up on sending it (#67 adds additional logging for this situation).

When this happens, one of the peers will never receive a termination message for that connection, making the underlying buffers to get stuck on `Read()` forever, hence causing goroutine and memory leaks.

## Solution

This PR adds a new message type to the protocol (`Resync`), whose payload contains a list of connection IDs. Similarly to how clients sends `Ping` control messages, `Resync` messages will periodically tell the receiving peer that any connection not contained in the provided list is no longer needed and can be pruned.

Small caveat: we cannot use Control messages for this purpose, since websocket set a limit of 125 bytes for their payload, which would impose a tight restriction on the number of connections.

## CheckList

- [X] Test

### CODE REVIEW

File: session_resync.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+			clientIDs = append(clientIDs[:i], clientIDs[i+1:]...) // IDs are unique, skip in next iteration
+			continue
+		}
+
+		// Connection no longer active in the client, close it server-side
+		conn := s.lockedRemoveConnection(id)
+		if conn != nil {
+			// TODO(alejandro.ruiz): the other end is gone, is this really needed?
+			// we could omit sending ERROR message to close the other end, but doTunnelClose must be executed (also metrics)
+			conn.tunnelClose(errCloseResync)
```
Comment (aruiz14) [inline session_resync.go:66]: Open question for reviewers:
`tunnelClose` does:
```
func (c *connection) tunnelClose(err error) {
	metrics.IncSMTotalRemoveConnectionsForWS(c.session.clientKey, c.addr.Network(), c.addr.String())
	c.writeErr(err)
	c.doTunnelClose(err)
}
```
1. Recording metrics is desired
2. `doTunnelClose` is required, so the associated `readBuffer` is closed.
3. However, `writeErr` will (try to) send a `Error` message in order to communicate this connection is going to be closed, but the other end does not know about it anymore.

Should I modify `tunnelClose` to not always call `writeErr`?

Comment (aruiz14) [reply session_resync.go]: Addressed in bfd6e57

File: connection.go [hunk 2]
```diff
@@ -46,6 +45,7 @@ func (c *connection) doTunnelClose(err error) {
 		return
 	}
 
+	metrics.IncSMTotalRemoveConnectionsForWS(c.session.clientKey, c.addr.Network(), c.addr.String())
```
Comment (aruiz14) [inline connection.go:48]: See https://github.com/rancher/remotedialer/pull/74/files/5e8dc8524011d90cf1cdcbd17713cb57f2d931a5#r1553397664 regarding this change

File: session_resync.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+// lockedResyncConnections closes any session connection that is not present in the IDs received from the client
+func (s *Session) lockedResyncConnections(payload []byte) error {
+	clientIDs, err := decodeConnectionIDs(payload)
+	if err != nil {
+		return fmt.Errorf("decoding resync connections payload: %w", err)
+	}
+	for _, id := range s.activeConnectionIDs() {
+		if i := sliceIndex(clientIDs, id); i >= 0 {
+			// Connection is still active, do nothing
+			clientIDs = append(clientIDs[:i], clientIDs[i+1:]...) // IDs are unique, skip in next iteration
```
Comment (moio) [inline session_resync.go:57]: Given we have two ordered lists (`activeConnectionIDs()` and `clientIDs`), how about using the classic merge-sort-like algorithm?

You keep two indices over the two lists, both starting at 0. Then you bump one or the other index depending on which list has the element with the greatest value. At each step you collect in a result list elements that are in the first but not in the second list.

File: session_resync.go (hunk 1 above)
Comment (moio) [inline session_resync.go]: IIUC:
 - `tunnelClose() -> doTunnelClose() -> c.buffer.close()` closes the read buffer allowing its goroutine to finish, which is important (arguably, the whole point of this exercise) - so that should happen
 - as you mentioned, it also updates metrics via `IncSMTotalRemoveConnectionsForWS`, which is important to maintain consistency
 - it also does `writeErr` which is bound to fail at the other end - I guess that should not happen

If that is correct, how about doing the equivalent of `tunnelClose` but without the `writeErr` part:

```
	metrics.IncSMTotalRemoveConnectionsForWS(c.session.clientKey, c.addr.Network(), c.addr.String())
	c.doTunnelClose(err)
```

Comment (moio) [reply session_resync.go]: OK you've already implemented something equivalent! 

🏁 🏁 🏁 🚗 🏁 🏁 🏁 

always faster than me 🦥

File: connection.go (hunk 2 above)
Comment (moio) [reply connection.go:48]: Won't this result in a double increment in `pipe`, which is currently the only other place calling `doTunnelClose()` directly?

File: message.go [hunk 4]
```diff
@@ -23,6 +23,7 @@ const (
 	RemoveClient
 	Pause
 	Resume
+	Resync
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply message.go:26]: Good point, I'll change it, thanks!

[the remark this answers is not in the document]
Comment (aruiz14) [reply message.go]: I've used `SyncConnections`

File: session_resync.go (hunk 3 above)
Comment (aruiz14) [reply session_resync.go]: Done at e7efd38, thanks!

File: connection.go (hunk 2 above)
Comment (moio) [reply connection.go:48]: Understood: the caller does that now.

Comment (moio) [reply connection.go:48]: IncSMTotalRemoveConnectionsForWS

**Review by moio — APPROVED** [review]
LGTM now, thanks

File: connection.go (hunk 2 above)
Comment (aruiz14) [reply connection.go:48]: FTR: before these changes, `doTunnelClose` is being called 3 times in total (2 from `pipe`, 1 from `clientDial` by deferring `conn.Close`, which then calls `conn.tunnelClose`). Of those 3 calls, only one of them will be effective, since there is a protection (checking `c.err != nil`) for the function being called before. Since I placed the `metrics` recording after this protection, it will keep being increased only in one of those calls.

File: session_sync.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	"errors"
+	"fmt"
+	"time"
+)
+
+var errCloseSyncConnections = errors.New("sync from client")
+
+// encodeConnectionIDs serializes a slice of connection IDs
+func encodeConnectionIDs(ids []int64) []byte {
+	payload := make([]byte, 0, 8*len(ids))
```
Comment (aruiz14) [inline session_sync.go:14]: nit-pick and open question for reviewers:

The size of the payload will be proportional to the number of active connections and depend on the type used for this value (`int64` at the moment, which needs 8 bytes to represent up to `9,22e18` (19 digits) positive values, since the counter starts at 0).
We could halve the payload size by using `uint32`, which would still leave us more than 4 billion unique ids to use, which I believe is more than enough for this purpose.
In fact, these ID are used in the context of a session (websocket multiplexing), so we could go even further and use `uint16`'s `65536` unique values.
The current implementation relies on `int64` providing an "almost infinite" number of IDs so they are not repeated, but we could also prevent that using a smaller size by ensuring a new connection's ID is not included in the connections map.

Finally, please beware this won't only affect the payload size, but also reduce the overall memory usage by remotedialer (although this may represent a tiny percentage from total Rancher usage).

Comment (tomleb) [reply session_sync.go:14]: Do we have a ballpark idea of how many active connections are opened in a typical case and in big clusters? Are we talking about 1000s, 10000s, more?

**Review by rmweir — CHANGES_REQUESTED** [review]

File: session_sync.go (hunk 5 above)
Comment (dvob) [reply session_sync.go:14]: @tomleb Currently on a rancher instance with around 3.4k connected downstream clusters I see 66.5k connections.
```
sum(session_server_total_add_connections - session_server_total_remove_connections)
```

The main question in general to me seems: What is the number of downstream cluster which Rancher should support and is remotedialer able to handle this amount of clusters/traffic?
Currently in our case we're not able to scale rancher to more than one pod as it seems that the remotedialer peer connections are not able to handle the load (see https://github.com/rancher/rancher/issues/45129). With a single pod we don't have peer connections and hence its working more or less.

Comment (dvob) [reply session_sync.go:14]: @moio thank you for the response.

>  I believe you are affected by the problem this PR is trying to fix

I didn't look into this PR in detail. But should the number of connections not grow steadily as they get lost?
Because in our case the connection count only goes up until all clusters are active. After that they stay on around 66.5k.
number_of_connections

Comment (dvob) [reply session_sync.go:14]: Its repeatedly ramping up because rancher is restarting all the time (see https://github.com/rancher/rancher/issues/45164). But during off-peak hours it usually stays steady on 66.5k for longer periods of time without restarting.

From my understanding the ramp up is because it takes some time until all controllers for the downstream clusters are started (user controllers?). After that all connected downstream clusters are in active state. Before they are in unavailable state. We see the same ramp up pattern for memory and network.
My assumtion is that during this startup phase the controllers are opening many connections because they are starting listeners (LIST WATCH)?

**Review by tomleb — COMMENTED** [review]
This seems to follow what we discussed during our meeting.

Given the potential deadlock mentioned in the comments below, I think this requires more testing. It should be tested in a custom Rancher build and it would be beneficial imo to have integration tests for this, though I understand that right now remotedialer doesn't have this in place.

File: session.go [hunk 6]
```diff
@@ -152,6 +166,13 @@ func (s *Session) serveMessage(ctx context.Context, reader io.Reader) error {
 		err := s.removeRemoteClient(message.address)
 		s.Unlock()
 		return err
+	} else if message.messageType == SyncConnections {
+		defer s.Unlock()
+		payload, err := io.ReadAll(message.body)
+		if err != nil {
+			return fmt.Errorf("reading message body: %w", err)
+		}
+		return s.lockedSyncConnections(payload)
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply session.go:175]: Thanks for the suggestion. I tried to not modify existing code too much, but happy to refactor this.
I've created a separate PR just for moving the code and making this serveMessage clearer:
https://github.com/rancher/remotedialer/pull/78

File: message.go [hunk 7]
```diff
@@ -23,6 +23,7 @@ const (
 	RemoveClient
 	Pause
 	Resume
+	SyncConnections
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply message.go:26]: Sure, will do! I also added comments to the rest of types as part of https://github.com/rancher/remotedialer/pull/78

File: session.go [hunk 8]
```diff
@@ -246,6 +266,26 @@ func (s *Session) closeConnection(connID int64, err error) {
...
+	s.Lock()
+	defer s.Unlock()
+
+	res := make([]int64, 0, len(s.conns))
+	for id := range s.conns {
+		res = append(res, id)
+	}
+	sort.Slice(res, func(i, j int) bool { return res[i] < res[j] })
+	return res
+}
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply session.go:287]: Thanks, did in a separete refactoring PR:
https://github.com/rancher/remotedialer/pull/78

File: session_sync.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+// lockedSyncConnections closes any session connection that is not present in the IDs received from the client
+// The session lock must be hold by the caller when calling this method
+func (s *Session) lockedSyncConnections(payload []byte) error {
+	clientIDs, err := decodeConnectionIDs(payload)
+	if err != nil {
+		return fmt.Errorf("decoding sync connections payload: %w", err)
+	}
+	serverIDs := s.activeConnectionIDs()
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply session_sync.go:55]: Nice catch, thank you. I took this into account when refactoring at https://github.com/rancher/remotedialer/pull/78, where I tried to remove the need to manage the lock from callers.
I've stacked this PR on top of that one and should be fixed not.

File: session_sync.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+		conn := s.lockedRemoveConnection(id)
+		if conn != nil {
+			// Using doTunnelClose directly instead of tunnelClose, omitting unnecessarily sending an Error message
+			conn.doTunnelClose(errCloseSyncConnections)
+		}
+	}
+	return nil
+}
+
+func removedFromSlice(a, b []int64) (res []int64) {
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply session_sync.go:67]: Tom is correct. Sorry if this was not straightforward, I've added a comment to try to make it clearer.
Please let me know if I should still change the name, thanks.

File: session_sync_test.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+				return
+			}
+		}
+	}()
+
+	for i := 1; i <= 4; i++ {
+		if err := session.sendPing(); err != nil {
+			t.Fatal(err)
+		}
+		time.Sleep(50 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply session_sync_test.go:114]: You are right, thanks! For the sake of the test, we don't even need to count them, so I've reworked it to use channels inspired by your example, thanks!

File: session_sync.go (hunk 5 above)
Comment (aruiz14) [reply session_sync.go:14]: > Say we'd use int64 then that'd be around 8KB sent every minute. Doesn't sound to bad to me, though maybe others disagree.

I agree, probably not worth it.

> If we plan on optimizing this in the future, it could help adding an "encoding type" byte to the sync connection payload so we can keep backwards compatible with previous version of remotedialer (eg: when doing a rolling upgrade). Another way would be to create a new SyncConnection message type, but the encoding type seems simpler to me.

That's a good idea. After thinking about it, maybe we don't even need to be this careful. If we ever changed it, all the remotedialer peers would be running the same version, since they would have the same Rancher image.
The only exception would be the upgrade time window, but we can basically ignore that, because all those connections are going to be dropped anyways.

---
In summary, I think there is no action needed here, so I'm going to resolve this conversation. Thanks for participating!

File: session_sync.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (nflynt) [reply session_sync.go:71]: If I'm following the logic correctly, the operation being performed is a **set difference*: we want all elements present in slice A, which are not present in slice B. Since the function will fail if either input slice is not sorted and the function signature doesn't guarantee this constraint, I would lean towards something like `differenceOfSortedSets(a, b []int64)`

My thinking is that using the name of the set operation should make the intended behavior easy to search, and documenting the sorting requirement in the name will make it harder to miss this constraint at call sites. (The sorting requirement permits an efficient single-pass algorithm, which makes it worth the added complexity.)

[the remark this answers is not in the document]
Comment (aruiz14) [reply session_sync.go]: Thanks, that makes sense and I've renamed the function accordingly.

**Review by moio — APPROVED** [review]

**Review by maxsokolovsky — CHANGES_REQUESTED** [review]
My comments are mostly nits at this point, since I am not yet familiar with this project. I may have to come back and think about the solution itself once I learn a bit more about remotedialer.

File: session_test.go [hunk 12]
```diff
@@ -75,3 +76,76 @@ func TestSession_sessionKeys(t *testing.T) {
 		t.Errorf("incorrect number of remote client keys after removal, got: %d, want %d", got, want)
 	}
 }
+
+func TestSession_activeConnectionIDs(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply session_test.go:84]: Thanks for the suggestion. Please note that, since the session is not shared between tests cases, they will all have their own lock though

File: session_sync.go [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+// sendSyncConnections sends a binary message of type SyncConnections, whose payload is a list of the active connection IDs for this session
+func (s *Session) sendSyncConnections() error {
+	_, err := s.writeMessage(time.Now().Add(SyncConnectionsTimeout), newSyncConnectionsMessage(s.activeConnectionIDs()))
+	return err
+}
+
+// closeMissingConnections closes any session connection that is not present in the IDs received from the client
+// The session lock must be hold by the caller when calling this method
+func (s *Session) closeStaleConnections(clientIDs []int64) {
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply session_sync.go]: Done in 14c3d27777ddef295ed1760351ec21cd173b7026

File: session_sync_test.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+		size int
+	}{
+		{0},
+		{1},
+		{2},
+		{10},
+		{100},
+		{1000},
+	}
+	for _, tt := range tests {
```
Comment (maxsokolovsky) [inline session_sync_test.go:30]: You'd need to capture the loop variable to avoid the problem with the same variable would be used in closures when multiple goroutines are present.

```suggestion
	for _, tt := range tests {
	      tt := tt
```

https://eli.thegreenplace.net/2019/go-internals-capturing-loop-variables-in-closures/

File: session_test.go [hunk 15]
```diff
@@ -75,3 +80,80 @@ func TestSession_sessionKeys(t *testing.T) {
...
+			name: "multiple connections",
+			conns: map[int64]*connection{
+				5:  nil,
+				20: nil,
+				3:  nil,
+			},
+			expected: []int64{3, 5, 20},
+		},
+	}
+	for _, tt := range tests {
```
Comment (maxsokolovsky) [inline session_test.go:113]:
```suggestion
	for _, tt := range tests {
	      tt := tt
```

File: session_sync_test.go [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+			expected: []int64{4, 5},
+		},
+		{
+			// Combined case
+			server:   []int64{1, 2, 3, 4, 5},
+			client:   []int64{3, 6},
+			expected: []int64{1, 2, 4, 5},
+		},
+	}
+	for x, tt := range tests {
```
Comment (maxsokolovsky) [inline session_sync_test.go:76]:
```suggestion
	for _, tt := range tests {
	      x := x; tt := tt
```

File: session_sync_test.go (hunk 14 above)
Comment (aruiz14) [reply session_sync_test.go:30]: Wow, I completely missed it, thanks. This caveat is changing in 1.22 though:
https://go.dev/blog/loopvar-preview
I have that version locally:
```
❯ go test -v -run 'Test_diffSortedSetsGetRemoved.*' ./
=== RUN   Test_diffSortedSetsGetRemoved
=== PAUSE Test_diffSortedSetsGetRemoved
=== CONT  Test_diffSortedSetsGetRemoved
=== RUN   Test_diffSortedSetsGetRemoved/case_0
=== PAUSE Test_diffSortedSetsGetRemoved/case_0
=== RUN   Test_diffSortedSetsGetRemoved/case_1
=== PAUSE Test_diffSortedSetsGetRemoved/case_1
=== RUN   Test_diffSortedSetsGetRemoved/case_2
=== PAUSE Test_diffSortedSetsGetRemoved/case_2
=== RUN   Test_diffSortedSetsGetRemoved/case_3
=== PAUSE Test_diffSortedSetsGetRemoved/case_3
=== CONT  Test_diffSortedSetsGetRemoved/case_0
=== CONT  Test_diffSortedSetsGetRemoved/case_2
=== CONT  Test_diffSortedSetsGetRemoved/case_1
=== CONT  Test_diffSortedSetsGetRemoved/case_3
--- PASS: Test_diffSortedSetsGetRemoved (0.00s)
    --- PASS: Test_diffSortedSetsGetRemoved/case_0 (0.00s)
    --- PASS: Test_diffSortedSetsGetRemoved/case_2 (0.00s)
    --- PASS: Test_diffSortedSetsGetRemoved/case_1 (0.00s)
    --- PASS: Test_diffSortedSetsGetRemoved/case_3 (0.00s)
PASS
ok  	github.com/rancher/remotedialer	0.209s
```
but it still not used in this project.

Having said that, I'm not sure it's worth the "complexity", just for running these tests in parallel. Locally for me, they used to always run under 1 second, and using `Parallel()` and `-p=10` or `20` does not have any noticeable impact.
I'm keen to revert that change and keep them simpler, since I believe these unit tests won't benefit much from running in parallel. What do you think?

Comment (maxsokolovsky) [reply session_sync_test.go:30]: Even in Go 1.22, this old behavior is observed. Basically, the last subtest will be run twice. So if the last subtest passes, it would appear as if the whole test passes, even if previous subtests fail. This behavior hides test failures.
In Go 1.22, you can forego the capturing of the loop variable and have the expected behavior if you use the experimental environment variable `GOEXPERIMENT=loopvar go test`. Maybe I am missing something, but it's not the default yet.

It's always good to run tests and subtests in parallel because Go would run parallel tests together with **other** parallel tests - that's the main benefit. It's not about running these individual tests in parallel in isolation. If your functions under test took more time (you can simulate this with `time.Sleep` calls), you'd see the difference in running them sequentially vs. in parallel vs. in parallel with parallel subtests.

Of course, tests and functions under test need to be eligible to be run in parallel, and that's possible when there is no global state shared among them. So running in parallel is a safe default unless there is a specific reason not to do this.

Comment (aruiz14) [reply session_sync_test.go:30]: Oh, you are right, the subtest name is created outside the closure, that's why it's correct.

OK, I'll keep the parallel marks, no strong opinion from my side.

Comment (aruiz14) [reply session_sync_test.go]: > Even in Go 1.22, this old behavior is observed.

From the link I posted:
> To ensure backwards compatibility with existing code, the new semantics will only apply in packages contained in modules that declare go 1.22 or later in their go.mod files.

Comment (maxsokolovsky) [reply session_sync_test.go]: Ah, you're right! So if the module declares that it uses 1.20, it doesn't matter that my local toolchain is 1.22, makes sense, thanks.

**Review by moio — APPROVED** [review]

**Comment by moio** [comment]
@maxsokolovsky do we have your final blessing to merge this, so that it has a chance to get into future 2.9.0 alphas?

**Review by maxsokolovsky — APPROVED** [review]
